<h1>Table of Contents<span class="tocSkip"></span></h1>
<div class="toc"><ul class="toc-item"><li><span><a href="#Customers-Dataset" data-toc-modified-id="Customers-Dataset-1"><span class="toc-item-num">1&nbsp;&nbsp;</span>Customers Dataset</a></span></li><li><span><a href="#Geolocation-Dataset" data-toc-modified-id="Geolocation-Dataset-2"><span class="toc-item-num">2&nbsp;&nbsp;</span>Geolocation Dataset</a></span></li><li><span><a href="#Order-Items" data-toc-modified-id="Order-Items-3"><span class="toc-item-num">3&nbsp;&nbsp;</span>Order Items</a></span></li><li><span><a href="#Order-Payments" data-toc-modified-id="Order-Payments-4"><span class="toc-item-num">4&nbsp;&nbsp;</span>Order Payments</a></span></li><li><span><a href="#Order-Reviews" data-toc-modified-id="Order-Reviews-5"><span class="toc-item-num">5&nbsp;&nbsp;</span>Order Reviews</a></span></li><li><span><a href="#Orders" data-toc-modified-id="Orders-6"><span class="toc-item-num">6&nbsp;&nbsp;</span>Orders</a></span></li><li><span><a href="#Products" data-toc-modified-id="Products-7"><span class="toc-item-num">7&nbsp;&nbsp;</span>Products</a></span></li><li><span><a href="#Sellers" data-toc-modified-id="Sellers-8"><span class="toc-item-num">8&nbsp;&nbsp;</span>Sellers</a></span></li><li><span><a href="#Product-Cat-Translation" data-toc-modified-id="Product-Cat-Translation-9"><span class="toc-item-num">9&nbsp;&nbsp;</span>Product Cat Translation</a></span></li><li><span><a href="#Testes-Join" data-toc-modified-id="Testes-Join-10"><span class="toc-item-num">10&nbsp;&nbsp;</span>Testes Join</a></span></li><li><span><a href="#Insights-EDA" data-toc-modified-id="Insights-EDA-11"><span class="toc-item-num">11&nbsp;&nbsp;</span>Insights EDA</a></span><ul class="toc-item"><li><span><a href="#Perguntas" data-toc-modified-id="Perguntas-11.1"><span class="toc-item-num">11.1&nbsp;&nbsp;</span>Perguntas</a></span></li></ul></li></ul></div>

Este notebook visa avaliar todos os datasets contidos no pacote `Brazilian e-commerce` do Kaggle. No total, são 9 datasets cada qual referente a um tema específico. Vamos explorar cada um deles para que, dessa forma, um problema de negócio possa ser definido e executado.

In [1]:
# Importando bibliotecas
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline

## Customers Dataset

In [4]:
# Lendo arquivo
df_customers = pd.read_csv('olist_customers_dataset.csv')
df_customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

In [3]:
# dimensões
df_customers.shape

(99441, 5)

In [4]:
# Dados nulos
df_customers.isnull().sum()

customer_id                 0
customer_unique_id          0
customer_zip_code_prefix    0
customer_city               0
customer_state              0
dtype: int64

In [5]:
# Tipos primitivos
df_customers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 5 columns):
customer_id                 99441 non-null object
customer_unique_id          99441 non-null object
customer_zip_code_prefix    99441 non-null int64
customer_city               99441 non-null object
customer_state              99441 non-null object
dtypes: int64(1), object(4)
memory usage: 3.8+ MB


In [6]:
# Quantos estados diferentes?
df_customers['customer_state'].value_counts()

SP    41746
RJ    12852
MG    11635
RS     5466
PR     5045
SC     3637
BA     3380
DF     2140
ES     2033
GO     2020
PE     1652
CE     1336
PA      975
MT      907
MA      747
MS      715
PB      536
PI      495
RN      485
AL      413
SE      350
TO      280
RO      253
AM      148
AC       81
AP       68
RR       46
Name: customer_state, dtype: int64

In [7]:
# Quantas cidades de SP?
df_sp = df_customers.query('customer_state == "SP"')
df_sp['customer_city'].value_counts()

sao paulo                     15540
campinas                       1444
guarulhos                      1189
sao bernardo do campo           938
santo andre                     796
osasco                          746
santos                          713
sao jose dos campos             691
sorocaba                        633
jundiai                         565
ribeirao preto                  510
barueri                         433
mogi das cruzes                 383
piracicaba                      369
sao jose do rio preto           335
carapicuiba                     328
maua                            323
praia grande                    311
taboao da serra                 296
diadema                         286
indaiatuba                      280
sao caetano do sul              277
bauru                           273
taubate                         270
cotia                           250
sao carlos                      236
suzano                          234
guaruja                     

In [8]:
# Vendo novamente
df_customers.head()

customer_id                customer_unique_id  \
0  06b8999e2fba1a1fbc88172c00ba8bc7  861eff4711a542e4b93843c6dd7febb0   
1  18955e83d337fd6b2def6b18a428ac77  290c77bc529b7ac935b93aa66c333dc3   
2  4e7b3e00288586ebd08712fdd0374a03  060e732b5b29e8181a18229c7b0b2b5e   
3  b2b6027bc5c5109e529d4dc6358b12c3  259dac757896d24d7702b9acbbff3f3c   
4  4f2d8ab171c80ec8364f7c12e35b23ad  345ecd01c38d18a9036ed96c73b8d066   

   customer_zip_code_prefix          customer_city customer_state  
0                     14409                 franca             SP  
1                      9790  sao bernardo do campo             SP  
2                      1151              sao paulo             SP  
3                      8775        mogi das cruzes             SP  
4                     13056               campinas             SP

O dataset `olist_customers_dataset.csv` contém 99441 entradas contendo informações sobre os clientes, como:
* id do cliente;
* id único do cliente;
* prefixo do cep;
* cidade;
* estado.

## Geolocation Dataset

In [14]:
# Lendo arquivo
df_geo = pd.read_csv('olist_geolocation_dataset.csv')

df_geo.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [15]:
df_geo.drop_duplicates(inplace=True)

In [16]:
df_geo.to_csv('olist_geolocation_dataset_without_duplicates.csv', index=False)

In [18]:
new_geo = pd.read_csv('olist_geolocation_dataset_without_duplicates.csv')
new_geo.head()

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
0                         1037       -23.545621       -46.639292   
1                         1046       -23.546081       -46.644820   
2                         1046       -23.546129       -46.642951   
3                         1041       -23.544392       -46.639499   
4                         1035       -23.541578       -46.641607   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [32]:
new_geo[new_geo.duplicated(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng'])]

geolocation_zip_code_prefix  geolocation_lat  geolocation_lng  \
128                            1043       -23.544735       -46.642523   
132                            1050       -23.548802       -46.642373   
134                            1024       -23.541390       -46.629899   
151                            1042       -23.544667       -46.640669   
160                            1020       -23.552496       -46.632060   
176                            1020       -23.552235       -46.628441   
179                            1009       -23.546341       -46.636239   
184                            1046       -23.546382       -46.642894   
186                            1011       -23.546690       -46.635447   
194                            1037       -23.543883       -46.638075   
205                            1046       -23.546081       -46.644820   
218                            1036       -23.542828       -46.639399   
223                            1017       -23.549813       -46.630350   
226                            1042       -23.545052       -46.640072   
231                            1013       -23.547686       -46.634140   
236                            1026       -23.539840       -46.630264   
239                            1009       -23.546935       -46.636588   
252                            1031       -23.542521       -46.635335   
254                            1046       -23.546758       -46.645771   
274                            1032       -23.541377       -46.635735   
290                            1027       -23.538015       -46.631947   
308                            1037       -23.545187       -46.637855   
317                            1050       -23.549854       -46.643139   
338                            1023       -23.541082       -46.630845   
340                            1046       -23.545355       -46.644014   
342                            1041       -23.543208       -46.640906   
346                            1050       -23.549774       -46.645599   
352                            1032       -23.539842       -46.635148   
361                            1017       -23.550011       -46.631166   
363                            1042       -23.544874       -46.640303   
...                             ...              ...              ...   
735592                        99200       -28.838720       -51.890546   
735616                        99250       -28.710500       -51.935310   
735636                        99200       -28.849039       -51.890516   
735679                        99200       -28.848193       -51.894094   
735711                        99200       -28.848218       -51.900289   
735738                        99200       -28.848124       -51.895351   
735758                        99200       -28.846448       -51.897337   
735772                        99200       -28.842018       -51.902937   
735800                        99200       -28.845494       -51.882686   
735855                        99250       -28.705634       -51.932766   
735911                        99320       -28.619204       -52.517827   
735992                        99320       -28.624838       -52.512802   
735996                        99315       -28.692173       -52.692514   
735997                        99380       -29.045658       -52.293821   
736206                        99440       -29.088281       -53.210607   
736297                        99470       -28.465992       -52.820491   
736331                        99440       -29.085803       -53.212804   
736354                        99470       -28.460437       -52.822999   
736357                        99440       -29.091132       -53.210837   
736365                        99470       -28.472175       -52.811217   
736846                        99525       -28.391557       -52.694514   
737802                        99855       -27.815066       -51.828740   
737969                        99870       -27.773141       -51.601

In [7]:
df_geo['geolocation_lat'] = df_geo['geolocation_lat'].astype(str).apply(lambda x: x.replace('.', ','))
df_geo['geolocation_lng'] = df_geo['geolocation_lng'].astype(str).apply(lambda x: x.replace('.', ','))
df_geo.head()

geolocation_zip_code_prefix      geolocation_lat      geolocation_lng  \
0                         1037  -23,545621281152677   -46,63929204800168   
1                         1046   -23,54608112703553   -46,64482029837157   
2                         1046   -23,54612896641469  -46,642951483611384   
3                         1041  -23,544392164868103   -46,63949930627844   
4                         1035   -23,54157796171149  -46,641607223296134   

  geolocation_city geolocation_state  
0        sao paulo                SP  
1        sao paulo                SP  
2        sao paulo                SP  
3        sao paulo                SP  
4        sao paulo                SP

In [13]:
df_geo.duplicated(['geolocation_zip_code_prefix', 'geolocation_lat', 'geolocation_lng', 'geolocation_city']).sum()

261833

In [10]:
# Dimensões
df_geo.shape

(1000163, 5)

In [11]:
df_geo['geolocation_state'].value_counts()

SP    404268
MG    126336
RJ    121169
RS     61851
PR     57859
SC     38328
BA     36045
GO     20139
ES     16748
PE     16432
DF     12986
MT     12031
CE     11674
PA     10853
MS     10431
MA      7853
PB      5538
RN      5041
PI      4549
AL      4183
TO      3576
SE      3563
RO      3478
AM      2432
AC      1301
AP       853
RR       646
Name: geolocation_state, dtype: int64

Aparentemente, este conjunto refere-se a dados geográficos de clientes que, em algum momento, realizaram compras online. Não foi identificado nenhuma chave estrangeira para linkar com outro dataset.

## Order Items

In [6]:
# Lendo dados
df_order_items = pd.read_csv('olist_order_items_dataset.csv')

df_order_items.head()

order_id  order_item_id  \
0  00010242fe8c5a6d1ba2dd792cb16214              1   
1  00018f77f2f0320c557190d7a144bdd3              1   
2  000229ec398224ef6ca0657da4fc703e              1   
3  00024acbcdf0a6daa1e931b038114c75              1   
4  00042b26cf59d7ce69dfabb4e55b4fd9              1   

                         product_id                         seller_id  \
0  4244733e06e7ecb4970a6e2683c13e61  48436dade18ac8b2bce089ec2a041202   
1  e5f2d52b802189ee658865ca93d83a8f  dd7ddc04e1b6c2c614352b383efe2d36   
2  c777355d18b72b67abbeef9df44fd0fd  5b51032eddd242adc84c38acab88f23d   
3  7634da152a4610f1595efa32f14722fc  9d7a1d34a5052409006425275ba1c2b4   
4  ac6c3623068f30de03045865e4e10089  df560393f3a51e74553ab94004ba5c87   

   shipping_limit_date   price  freight_value  
0  2017-09-19 09:45:35   58.90          13.29  
1  2017-05-03 11:05:13  239.90          19.93  
2  2018-01-18 14:48:30  199.00          17.87  
3  2018-08-15 10:10:18   12.99          12.79  
4  2017-02-13 13:57:51  199.90          18.14

In [7]:
df_order_items.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 112650 entries, 0 to 112649
Data columns (total 7 columns):
order_id               112650 non-null object
order_item_id          112650 non-null int64
product_id             112650 non-null object
seller_id              112650 non-null object
shipping_limit_date    112650 non-null object
price                  112650 non-null float64
freight_value          112650 non-null float64
dtypes: float64(2), int64(1), object(4)
memory usage: 6.0+ MB


In [8]:
df_order_items.query('order_id == "b2eda1812564491a3461f7e08104ed8f"')

order_id  order_item_id  \
78650  b2eda1812564491a3461f7e08104ed8f              1   
78651  b2eda1812564491a3461f7e08104ed8f              2   
78652  b2eda1812564491a3461f7e08104ed8f              3   
78653  b2eda1812564491a3461f7e08104ed8f              4   

                             product_id                         seller_id  \
78650  cdb2ede3e6744ad7011a7305b95442a1  ad75519a07f039025e83795a4418470f   
78651  cdb2ede3e6744ad7011a7305b95442a1  ad75519a07f039025e83795a4418470f   
78652  cdb2ede3e6744ad7011a7305b95442a1  ad75519a07f039025e83795a4418470f   
78653  cdb2ede3e6744ad7011a7305b95442a1  ad75519a07f039025e83795a4418470f   

       shipping_limit_date  price  freight_value  
78650  2018-08-17 16:24:08  31.62            8.0  
78651  2018-08-17 16:24:08  31.62            8.0  
78652  2018-08-17 16:24:08  31.62            8.0  
78653  2018-08-17 16:24:08  31.62            8.0

In [13]:
# Dimensões
df_order_items.shape

(112650, 7)

In [14]:
df_order_items['order_item_id'].value_counts()

1     98666
2      9803
3      2287
4       965
5       460
6       256
7        58
8        36
9        28
10       25
11       17
12       13
13        8
14        7
15        5
20        3
16        3
17        3
18        3
19        3
21        1
Name: order_item_id, dtype: int64

Este dataset contém informações sobre os itens de cada pedido, como:
* id do pedido;
* id do item do pedido (provavelmente quantidade);
* id do produto;
* id do vendedor;
* data limite para entrega;
* preço;
* valor do frete.

## Order Payments

In [9]:
# Importando dados
df_order_payments = pd.read_csv('olist_order_payments_dataset.csv')

df_order_payments.head()

order_id  payment_sequential payment_type  \
0  b81ef226f3fe1789b1e8b2acac839d17                   1  credit_card   
1  a9810da82917af2d9aefd1278f1dcfa0                   1  credit_card   
2  25e8ea4e93396b6fa0d3dd708e76c1bd                   1  credit_card   
3  ba78997921bbcdc1373bb41e913ab953                   1  credit_card   
4  42fdf880ba16b47b59251dd489d4441a                   1  credit_card   

   payment_installments  payment_value  
0                     8          99.33  
1                     1          24.39  
2                     1          65.71  
3                     8         107.78  
4                     2         128.45

In [10]:
df_order_payments.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 103886 entries, 0 to 103885
Data columns (total 5 columns):
order_id                103886 non-null object
payment_sequential      103886 non-null int64
payment_type            103886 non-null object
payment_installments    103886 non-null int64
payment_value           103886 non-null float64
dtypes: float64(1), int64(2), object(2)
memory usage: 4.0+ MB


In [16]:
df_order_payments.shape

(103886, 5)

Aqui podemos linkar **order_id**

In [17]:
df_order_payments['payment_sequential'].value_counts()

1     99360
2      3039
3       581
4       278
5       170
6       118
7        82
8        54
9        43
10       34
11       29
12       21
13       13
14       10
15        8
16        6
17        6
18        6
19        6
20        4
21        4
22        3
23        2
24        2
25        2
26        2
28        1
27        1
29        1
Name: payment_sequential, dtype: int64

In [18]:
df_order_payments['payment_type'].value_counts()

credit_card    76795
boleto         19784
voucher         5775
debit_card      1529
not_defined        3
Name: payment_type, dtype: int64

In [19]:
df_order_payments['payment_installments'].value_counts()

1     52546
2     12413
3     10461
4      7098
10     5328
5      5239
8      4268
6      3920
7      1626
9       644
12      133
15       74
18       27
11       23
24       18
20       17
13       16
14       15
17        8
16        5
21        3
0         2
23        1
22        1
Name: payment_installments, dtype: int64

## Order Reviews

In [11]:
# Lendo
df_order_reviews = pd.read_csv('olist_order_reviews_dataset.csv')

df_order_reviews.head()

review_id                          order_id  \
0  7bc2406110b926393aa56f80a40eba40  73fc7af87114b39712e6da79b0a377eb   
1  80e641a11e56f04c1ad469d5645fdfde  a548910a1c6147796b98fdf73dbeba33   
2  228ce5500dc1d8e020d8d1322874b6f0  f9e4b658b201a9f2ecdecbb34bed034b   
3  e64fb393e7b32834bb789ff8bb30750e  658677c97b385a9be170737859d3511b   
4  f7c4243c7fe1938f181bec41a392bdeb  8e6bfb81e283fa7e4f11123a3fb894f1   

   review_score review_comment_title  \
0             4                  NaN   
1             5                  NaN   
2             5                  NaN   
3             5                  NaN   
4             5                  NaN   

                              review_comment_message review_creation_date  \
0                                                NaN  2018-01-18 00:00:00   
1                                                NaN  2018-03-10 00:00:00   
2                                                NaN  2018-02-17 00:00:00   
3              Recebi bem antes do prazo estipulado.  2017-04-21 00:00:00   
4  Parabéns lojas lannister adorei comprar pela I...  2018-03-01 00:00:00   

  review_answer_timestamp  
0     2018-01-18 21:46:59  
1     2018-03-11 03:05:13  
2     2018-02-18 14:36:24  
3     2017-04-21 22:02:06  
4     2018-03-02 10:26:53

In [13]:
df_order_reviews.to_csv('olist_order_reviews_edited.csv', sep=';', index=False)

In [12]:
df_order_reviews.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 100000 entries, 0 to 99999
Data columns (total 7 columns):
review_id                  100000 non-null object
order_id                   100000 non-null object
review_score               100000 non-null int64
review_comment_title       11715 non-null object
review_comment_message     41753 non-null object
review_creation_date       100000 non-null object
review_answer_timestamp    100000 non-null object
dtypes: int64(1), object(6)
memory usage: 5.3+ MB


In [21]:
df_order_reviews['review_comment_message'].notnull().sum()

41753

In [22]:
df_order_reviews.shape

(100000, 7)

## Orders

In [2]:
# Lendo
df_orders = pd.read_csv('olist_orders_dataset.csv')

df_orders.head()















order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
2  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   
3  949d5b44dbf5de918fe9c16f97b45f8a  f88197465ea7920adcdbec7375364d82   
4  ad21c59c0840e6cb83a9ceb5573f8159  8ab97904e6daea8866dbdbc4fb7aad2c   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
2    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   
3    delivered      2017-11-18 19:28:06  2017-11-18 19:45:59   
4    delivered      2018-02-13 21:18:39  2018-02-13 22:20:29   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2018-07-26 14:31:00           2018-08-07 15:27:45   
2          2018-08-08 13:50:00           2018-08-17 18:06:29   
3          2017-11-22 13:39:59           2017-12-02 00:28:42   
4          2018-02-14 19:46:34           2018-02-16 18:17:02   

  order_estimated_delivery_date  
0           2017-10-18 00:00:00  
1           2018-08-13 00:00:00  
2           2018-09-04 00:00:00  
3           2017-12-15 00:00:00  
4           2018-02-26 00:00:00

In [3]:
df_orders.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 99441 entries, 0 to 99440
Data columns (total 8 columns):
order_id                         99441 non-null object
customer_id                      99441 non-null object
order_status                     99441 non-null object
order_purchase_timestamp         99441 non-null object
order_approved_at                99281 non-null object
order_delivered_carrier_date     97658 non-null object
order_delivered_customer_date    96476 non-null object
order_estimated_delivery_date    99441 non-null object
dtypes: object(8)
memory usage: 6.1+ MB


In [24]:
df_orders.shape

(99441, 8)

In [25]:
df_orders['order_status'].value_counts()

delivered      96478
shipped         1107
canceled         625
unavailable      609
invoiced         314
processing       301
created            5
approved           2
Name: order_status, dtype: int64

## Products

In [2]:
# Lendo
df_prod = pd.read_csv('olist_products_dataset.csv')

df_prod.head()

product_id  product_category_name  \
0  1e9e8ef04dbcff4541ed26657ea517e5             perfumaria   
1  3aa071139cb16b67ca9e5dea641aaa2f                  artes   
2  96bd76ec8810374ed1b65e291975717f          esporte_lazer   
3  cef67bcfe19066a932b7673e239eb23d                  bebes   
4  9dc1a7de274444849c219cff195d0b71  utilidades_domesticas   

   product_name_lenght  product_description_lenght  product_photos_qty  \
0                 40.0                       287.0                 1.0   
1                 44.0                       276.0                 1.0   
2                 46.0                       250.0                 1.0   
3                 27.0                       261.0                 1.0   
4                 37.0                       402.0                 4.0   

   product_weight_g  product_length_cm  product_height_cm  product_width_cm  
0             225.0               16.0               10.0              14.0  
1            1000.0               30.0               18.0              20.0  
2             154.0               18.0                9.0              15.0  
3             371.0               26.0                4.0              26.0  
4             625.0               20.0               17.0              13.0

In [5]:
id_teste = "60d1e2123927bc621354b0ab11cc8fd5"
df_prod.query('product_id == @id_teste')

product_id product_category_name  \
3699  60d1e2123927bc621354b0ab11cc8fd5                   NaN   

      product_name_lenght  product_description_lenght  product_photos_qty  \
3699                  NaN                         NaN                 NaN   

      product_weight_g  product_length_cm  product_height_cm  product_width_cm  
3699             300.0               16.0               19.0              16.0

In [27]:
df_prod.shape

(32951, 9)

In [28]:
df_prod['product_category_name'].value_counts()

cama_mesa_banho                                  3029
esporte_lazer                                    2867
moveis_decoracao                                 2657
beleza_saude                                     2444
utilidades_domesticas                            2335
automotivo                                       1900
informatica_acessorios                           1639
brinquedos                                       1411
relogios_presentes                               1329
telefonia                                        1134
bebes                                             919
perfumaria                                        868
fashion_bolsas_e_acessorios                       849
papelaria                                         849
cool_stuff                                        789
ferramentas_jardim                                753
pet_shop                                          719
eletronicos                                       517
construcao_ferramentas_const

## Sellers

In [29]:
# Lendo
df_sellers = pd.read_csv('olist_sellers_dataset.csv')

df_sellers.head()

seller_id  seller_zip_code_prefix  \
0  3442f8959a84dea7ee197c632cb2df15                   13023   
1  d1b65fc7debc3361ea86b5f14c68d2e2                   13844   
2  ce3ad9de960102d0677a81f5d0bb7b2d                   20031   
3  c0f3eea2e14555b6faeea3dd58c1b1c3                    4195   
4  51a04a8a6bdcb23deccc82b0b80742cf                   12914   

         seller_city seller_state  
0           campinas           SP  
1         mogi guacu           SP  
2     rio de janeiro           RJ  
3          sao paulo           SP  
4  braganca paulista           SP

In [30]:
df_sellers.shape

(3095, 4)

## Product Cat Translation

In [31]:
# Lendo
df_prod_translation = pd.read_csv('product_category_name_translation.csv')

df_prod_translation.head()

product_category_name product_category_name_english
0            beleza_saude                 health_beauty
1  informatica_acessorios         computers_accessories
2              automotivo                          auto
3         cama_mesa_banho                bed_bath_table
4        moveis_decoracao               furniture_decor

In [32]:
df_prod_translation.shape

(71, 2)

## Testes Join

Temos os seguintes datasets:
    
    df_customers
    df_geo
    df_order_items
    df_order_payments
    df_order_reviews
    df_orders
    df_prod
    df_sellers    

In [49]:
# Unindo tudo o que é possível
print(f'Dimensões de df_orders: {df_orders.shape}')
print(f'Dimensões de df_order_payments: {df_order_payments.shape}')
df_orders_and_payments = df_orders.merge(df_order_payments, on='order_id', how='inner')

print(f'Dimensão da união: {df_orders_and_payments.shape}')

# Primeiras linhas
df_orders_and_payments.head()

Dimensões de df_orders: (99441, 8)
Dimensões de df_order_payments: (103886, 5)
Dimensão da união: (103886, 12)


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  payment_sequential payment_type  \
0           2017-10-18 00:00:00                   1  credit_card   
1           2017-10-18 00:00:00                   3      voucher   
2           2017-10-18 00:00:00                   2      voucher   
3           2018-08-13 00:00:00                   1       boleto   
4           2018-09-04 00:00:00                   1  credit_card   

   payment_installments  payment_value  
0                     1          18.12  
1                     1           2.00  
2                     1          18.59  
3                     1         141.46  
4                     3         179.12

In [50]:
# Verificando
df_orders_and_payments.isnull().any()

order_id                         False
customer_id                      False
order_status                     False
order_purchase_timestamp         False
order_approved_at                 True
order_delivered_carrier_date      True
order_delivered_customer_date     True
order_estimated_delivery_date    False
payment_sequential               False
payment_type                     False
payment_installments             False
payment_value                    False
dtype: bool

In [55]:
# Verificando entradas duplicadas
df_orders_and_payments[df_orders_and_payments.duplicated('order_id', keep=False) == True]

order_id                       customer_id  \
0       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2       e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
11      e69bfb5eb88e0ed6a785585b27e16dbf  31ad1d1b63eb9962463f764d4e6e0c9d   
12      e69bfb5eb88e0ed6a785585b27e16dbf  31ad1d1b63eb9962463f764d4e6e0c9d   
22      83018ec114eee8641c97e08f7b4e926f  7f8c8b9c2ae27bf3300f670c3d478be8   
23      83018ec114eee8641c97e08f7b4e926f  7f8c8b9c2ae27bf3300f670c3d478be8   
24      83018ec114eee8641c97e08f7b4e926f  7f8c8b9c2ae27bf3300f670c3d478be8   
82      634e8f4c0f6744a626f77f39770ac6aa  05e996469a2bf9559c7122b87e156724   
83      634e8f4c0f6744a626f77f39770ac6aa  05e996469a2bf9559c7122b87e156724   
104     f7959f8385f34c4f645327465a1c9fc4  0bf19317b1830a69e55b40710576aa7a   
105     f7959f8385f34c4f645327465a1c9fc4  0bf19317b1830a69e55b40710576aa7a   
170     b2bd1f09c1b8a4a36940f6f6e49d5dec  8ed14e450a6268ec13f325e7c2eafeed   
171     b2bd1f09c1b8a4a36940f6f6e49d5dec  8ed14e450a6268ec13f325e7c2eafeed   
172     b2bd1f09c1b8a4a36940f6f6e49d5dec  8ed14e450a6268ec13f325e7c2eafeed   
217     a03ad7057fae696dd18f8967826c209f  7df71d1652cfba8b2a70cf665c960f74   
218     a03ad7057fae696dd18f8967826c209f  7df71d1652cfba8b2a70cf665c960f74   
260     fd9b8ed51e5ffedf7a56bea36cc5b6f9  0385c03f2b10a0a50ee72d8e3649a4b3   
261     fd9b8ed51e5ffedf7a56bea36cc5b6f9  0385c03f2b10a0a50ee72d8e3649a4b3   
262     bb4f0c21ae6014e40ef007ff6fbcbeb9  94f7f1cb719dd707fd437e918d282537   
263     bb4f0c21ae6014e40ef007ff6fbcbeb9  94f7f1cb719dd707fd437e918d282537   
264     f31535f21d145b2345e2bf7f09d62322  ed532487db04478dfba09d112649fc13   
265     f31535f21d145b2345e2bf7f09d62322  ed532487db04478dfba09d112649fc13   
306     1d251ab94983c4adb11e4b168abb1439  230ab395e91b2421bd00ea45b60b23cb   
307     1d251ab94983c4adb11e4b168abb1439  230ab395e91b2421bd00ea45b60b23cb   
308     1d251ab94983c4adb11e4b168abb1439  230ab395e91b2421bd00ea45b60b23cb   
333     dc400373e624e9b0435847a79813c658  4a00f00d951c167739888a1f16c56e8e   
334     dc400373e624e9b0435847a79813c658  4a00f00d951c167739888a1f16c56e8e   
352     d7f4e2f755cf2f40aa7c39804cfe01c4  18ad705482141f60ad05945506ef2b9e   
353     d7f4e2f755cf2f40aa7c39804cfe01c4  18ad705482141f60ad05945506ef2b9e   
...                                  ...                               ...   
103542  274a7f7e4f1c17b7434a830e9b8759b1  670af30ca5b8c20878fecdafa5ee01b9   
103543  274a7f7e4f1c17b7434a830e9b8759b1  670af30ca5b8c20878fecdafa5ee01b9   
103559  b93b079468bdf6a20fd54cf45d1e03b6  e570debfd69192cdd2e846ae00869a33   
103560  b93b079468bdf6a20fd54cf45d1e03b6  e570debfd69192cdd2e846ae00869a33   
103564  e9d424615d8805614e254bf5c170bef8  102a23c6841127cf625296ef67d5f79c   
103565  e9d424615d8805614e254bf5c170bef8  102a23c6841127cf625296ef67d5f79c   
103566  e9d424615d8805614e254bf5c170bef8  102a23c6841127cf625296ef67d5f79c   
103579  e32ebf133fb19cafd04315ff6d108047  b65a67a6cb9f36ff9379fcc0c88cccb8   
103580  e32ebf133fb19cafd04315ff6d108047  b65a67a6cb9f36ff9379fcc0c88cccb8   
103620  8bbfd0be5d89a92d4bf214baac246fea  85e0c3aa291ef2ca5f2b48d2aa1c9904   
103621  8bbfd0be5d89a92d4bf214baac246fea  85e0c3aa291ef2ca5f2b48d2aa1c9904   
103637  74d882fe8bdccca5952fbe23321ecfa7  784ea06d707eeb460893a573fa879dc2   
103638  74d882fe8bdccca5952fbe23321ecfa7  784ea06d707eeb460893a573fa879dc2   
103668  304bf410920449182fb3294bc7405597  19aa722bdebc8edaf921d50fd35916d1   
103669  304bf410920449182fb3294bc7405597  19aa722bdebc8edaf921d50fd35916d1   
103670  1c7c8ab324927f5b6e52338bd110bef4  95b1767df51d6fa09402267d8de314df   
103671  1c7c8ab324927f5b6e52338bd110bef4  95b1767df51d6fa09402267d8de314df   
103675  8508c8962494e9606eb5eacaa61c22e2  98058e1fcf55c882ba711cfb5083bff5   
103676  8508c8962494e9606eb5eacaa61c22e2  98058e1fcf55c882ba711cfb5083bff5   
103738  08d138a502fa05af63ac4005311484f9  bb3ef6ad2a44

order_ids duplicados indicam diferentes meios de pagamentos escolhidos por clientes para uma mesmo compra (ex, credit card, voucher, ...). As diferenças encontram-se nas colunas `payment_sequential` e `payment_type`.

In [57]:
# Unindo customers
print(f'Dimensões de df_orders: {df_orders.shape}')
print(f'Dimensões de df_order_payments: {df_order_payments.shape}')
print(f'Dimensões de df_customers: {df_customers.shape}')

df_ord_pay_cust = df_orders_and_payments.merge(df_customers, on='customer_id', how='inner')
print(f'\nDimensão da união orders and payments: {df_orders_and_payments.shape}')
print(f'Dimensão da união ord pay and costumers: {df_ord_pay_cust.shape}')

# Primeiras linhas
df_ord_pay_cust.head()

Dimensões de df_orders: (99441, 8)
Dimensões de df_order_payments: (103886, 5)
Dimensões de df_customers: (99441, 5)

Dimensão da união orders and payments: (103886, 12)
Dimensão da união ord pay and costumers: (103886, 16)


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  payment_sequential payment_type  \
0           2017-10-18 00:00:00                   1  credit_card   
1           2017-10-18 00:00:00                   3      voucher   
2           2017-10-18 00:00:00                   2      voucher   
3           2018-08-13 00:00:00                   1       boleto   
4           2018-09-04 00:00:00                   1  credit_card   

   payment_installments  payment_value                customer_unique_id  \
0                     1          18.12  7c396fd4830fd04220f754e42b4e5bff   
1                     1           2.00  7c396fd4830fd04220f754e42b4e5bff   
2                     1          18.59  7c396fd4830fd04220f754e42b4e5bff   
3                     1         141.46  af07308b275d755c9edb36a90c618231   
4                     3         179.12  3a653a41f6f9fc3d2a113cf8398680e8   

   customer_zip_code_prefix customer_city customer_state  
0                      3149     sao paulo             SP  
1                      3149     sao paulo             SP  
2                      3149     sao paulo             SP  
3                     47813     barreiras             BA  
4                     75265    vianopolis             GO

Para análises futuras, não é interessante manter customers dentro deste dataset, visto que, como os pagamentos diferentes geram ids repetidos, contabilizações erradas podem ser realizadas, por exemplo, no levantamento de quantas compra foram realizadas em SP (a nao ser que seja realizado um Group By).

In [62]:
# Juntando review
print(f'Dimensões de df_orders: {df_orders.shape}')
print(f'Dimensões de df_order_payments: {df_order_payments.shape}')
print(f'Dimensões de df_customers: {df_customers.shape}')
print(f'Dimensões de df_reviews: {df_order_reviews.shape}')

df_ord_pay_cust_rev = df_ord_pay_cust.merge(df_order_reviews, on='order_id', how='inner')
print(f'\nDimensão da união orders and payments: {df_orders_and_payments.shape}')
print(f'Dimensão da união ord pay and costumers: {df_ord_pay_cust.shape}')
print(f'Dimensão da união ord pay cust review: {df_ord_pay_cust_rev.shape}')

# Primeiras linhas
df_ord_pay_cust_rev.head()

Dimensões de df_orders: (99441, 8)
Dimensões de df_order_payments: (103886, 5)
Dimensões de df_customers: (99441, 5)
Dimensões de df_reviews: (100000, 7)

Dimensão da união orders and payments: (103886, 12)
Dimensão da união ord pay and costumers: (103886, 16)
Dimensão da união ord pay cust review: (104485, 22)


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  payment_sequential payment_type  \
0           2017-10-18 00:00:00                   1  credit_card   
1           2017-10-18 00:00:00                   3      voucher   
2           2017-10-18 00:00:00                   2      voucher   
3           2018-08-13 00:00:00                   1       boleto   
4           2018-09-04 00:00:00                   1  credit_card   

            ...                          customer_unique_id  \
0           ...            7c396fd4830fd04220f754e42b4e5bff   
1           ...            7c396fd4830fd04220f754e42b4e5bff   
2           ...            7c396fd4830fd04220f754e42b4e5bff   
3           ...            af07308b275d755c9edb36a90c618231   
4           ...            3a653a41f6f9fc3d2a113cf8398680e8   

   customer_zip_code_prefix customer_city  customer_state  \
0                      3149     sao paulo              SP   
1                      3149     sao paulo              SP   
2                      3149     sao paulo              SP   
3                     47813     barreiras              BA   
4                     75265    vianopolis              GO   

                          review_id review_score review_comment_title  \
0  a54f0611adc9ed256b57ede6b6eb5114            4                  NaN   
1  a54f0611adc9ed256b57ede6b6eb5114            4                  NaN   
2  a54f0611adc9ed256b57ede6b6eb5114            4                  NaN   
3  8d5266042046a06655c8db133d120ba5            4     Muito boa a loja   
4  e73b67b67587f7644d5bd1a52deb1b01            5                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
2  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
3                               Muito bom o produto.  2018-08-08 00:00:00   
4                                                NaN  2018-08-18 00:00:00   

  review_answer_timestamp  
0     2017-10-12 03:43:48  
1     2017-10-12 03:43:48  
2     2017-10-12 03:43:48  
3     2018-08-08 18:37:50  
4     2018-08-22 19:07:58  

[5 rows x 22 columns]

In [63]:
# Unindo itens
# Juntando review
print(f'Dimensões de df_orders: {df_orders.shape}')
print(f'Dimensões de df_order_payments: {df_order_payments.shape}')
print(f'Dimensões de df_customers: {df_customers.shape}')
print(f'Dimensões de df_reviews: {df_order_reviews.shape}')
print(f'Dimensões de df_order_itens: {df_order_items.shape}')

df_all = df_ord_pay_cust_rev.merge(df_order_items, on='order_id', how='inner')
print(f'\nDimensão da união orders and payments: {df_orders_and_payments.shape}')
print(f'Dimensão da união ord pay and costumers: {df_ord_pay_cust.shape}')
print(f'Dimensão da união ord pay cust review: {df_ord_pay_cust_rev.shape}')
print(f'Dimensão de tudo unido: {df_all.shape}')

# Primeiras linhas
df_all.head()

Dimensões de df_orders: (99441, 8)
Dimensões de df_order_payments: (103886, 5)
Dimensões de df_customers: (99441, 5)
Dimensões de df_reviews: (100000, 7)
Dimensões de df_order_itens: (112650, 7)

Dimensão da união orders and payments: (103886, 12)
Dimensão da união ord pay and costumers: (103886, 16)
Dimensão da união ord pay cust review: (104485, 22)
Dimensão de tudo unido: (118315, 28)


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  53cdb2fc8bc7dce0b6741e2150273451  b0830fb4747a6c6d20dea0b8c802d7ef   
4  47770eb9100c2d0c44946d9cf07ec65d  41ce2a54c0b03bf3443c3d931a367089   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2018-07-24 20:41:37  2018-07-26 03:24:27   
4    delivered      2018-08-08 08:38:49  2018-08-08 08:55:23   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2018-07-26 14:31:00           2018-08-07 15:27:45   
4          2018-08-08 13:50:00           2018-08-17 18:06:29   

  order_estimated_delivery_date  payment_sequential payment_type  \
0           2017-10-18 00:00:00                   1  credit_card   
1           2017-10-18 00:00:00                   3      voucher   
2           2017-10-18 00:00:00                   2      voucher   
3           2018-08-13 00:00:00                   1       boleto   
4           2018-09-04 00:00:00                   1  credit_card   

       ...       review_comment_title  \
0      ...                        NaN   
1      ...                        NaN   
2      ...                        NaN   
3      ...           Muito boa a loja   
4      ...                        NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
2  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
3                               Muito bom o produto.  2018-08-08 00:00:00   
4                                                NaN  2018-08-18 00:00:00   

   review_answer_timestamp order_item_id                        product_id  \
0      2017-10-12 03:43:48             1  87285b34884572647811a353c7ac498a   
1      2017-10-12 03:43:48             1  87285b34884572647811a353c7ac498a   
2      2017-10-12 03:43:48             1  87285b34884572647811a353c7ac498a   
3      2018-08-08 18:37:50             1  595fac2a385ac33a80bd5114aec74eb8   
4      2018-08-22 19:07:58             1  aa4383b373c6aca5d8797843e5594415   

                          seller_id  shipping_limit_date   price freight_value  
0  3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15   29.99          8.72  
1  3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15   29.99          8.72  
2  3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15   29.99          8.72  
3  289cdb325fb7e7f891c38608bf9e0962  2018-07-30 03:24:27  118.70         22.76  
4  4869f7a5dfa277a7dca6462dcf3b52b2  2018-08-13 08:55:23  159.90         19.22  

[5 rows x 28 columns]

In [64]:
# Unindo sellers
print(f'Dimensões de df_orders: {df_orders.shape}')
print(f'Dimensões de df_order_payments: {df_order_payments.shape}')
print(f'Dimensões de df_customers: {df_customers.shape}')
print(f'Dimensões de df_reviews: {df_order_reviews.shape}')
print(f'Dimensões de df_order_itens: {df_order_items.shape}')
print(f'Dimensão de df_sellers: {df_sellers.shape}')

df_all = df_all.merge(df_sellers, on='seller_id', how='inner')
print(f'\nDimensão da união orders and payments: {df_orders_and_payments.shape}')
print(f'Dimensão da união ord pay and costumers: {df_ord_pay_cust.shape}')
print(f'Dimensão da união ord pay cust review: {df_ord_pay_cust_rev.shape}')
print(f'Dimensão de tudo unido: {df_all.shape}')

# Primeiras linhas
df_all.head()

Dimensões de df_orders: (99441, 8)
Dimensões de df_order_payments: (103886, 5)
Dimensões de df_customers: (99441, 5)
Dimensões de df_reviews: (100000, 7)
Dimensões de df_order_itens: (112650, 7)
Dimensão de df_sellers: (3095, 4)

Dimensão da união orders and payments: (103886, 12)
Dimensão da união ord pay and costumers: (103886, 16)
Dimensão da união ord pay cust review: (104485, 22)
Dimensão de tudo unido: (118315, 31)


order_id                       customer_id  \
0  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
1  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
2  e481f51cbdc54678b7cc49136f2d6af7  9ef432eb6251297304e76186b10a928d   
3  8736140c61ea584cb4250074756d8f3b  ab8844663ae049fda8baf15fc928f47f   
4  a0151737f2f0c6c0a5fd69d45f66ceea  fc2697314ab7fbeda62bb6f1afa4efcd   

  order_status order_purchase_timestamp    order_approved_at  \
0    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
1    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
2    delivered      2017-10-02 10:56:33  2017-10-02 11:07:15   
3    delivered      2017-08-10 13:35:55  2017-08-10 13:50:09   
4    delivered      2017-07-12 14:11:58  2017-07-12 14:30:18   

  order_delivered_carrier_date order_delivered_customer_date  \
0          2017-10-04 19:55:00           2017-10-10 21:25:13   
1          2017-10-04 19:55:00           2017-10-10 21:25:13   
2          2017-10-04 19:55:00           2017-10-10 21:25:13   
3          2017-08-11 13:52:35           2017-08-16 19:03:36   
4          2017-07-13 12:47:05           2017-07-18 16:24:42   

  order_estimated_delivery_date  payment_sequential payment_type     ...       \
0           2017-10-18 00:00:00                   1  credit_card     ...        
1           2017-10-18 00:00:00                   3      voucher     ...        
2           2017-10-18 00:00:00                   2      voucher     ...        
3           2017-08-23 00:00:00                   1  credit_card     ...        
4           2017-07-25 00:00:00                   1  credit_card     ...        

   review_answer_timestamp  order_item_id                        product_id  \
0      2017-10-12 03:43:48              1  87285b34884572647811a353c7ac498a   
1      2017-10-12 03:43:48              1  87285b34884572647811a353c7ac498a   
2      2017-10-12 03:43:48              1  87285b34884572647811a353c7ac498a   
3      2017-08-21 12:43:27              1  b00a32a0b42fd65efb58a5822009f629   
4      2017-07-24 13:59:59              1  725cbfcaff95a4d43742fdf13cf43c75   

                          seller_id  shipping_limit_date  price freight_value  \
0  3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15  29.99          8.72   
1  3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15  29.99          8.72   
2  3504c0cb71d7fa48d967e0e4c94d59d9  2017-10-06 11:07:15  29.99          8.72   
3  3504c0cb71d7fa48d967e0e4c94d59d9  2017-08-16 13:50:09  75.90          7.79   
4  3504c0cb71d7fa48d967e0e4c94d59d9  2017-07-18 14:30:18  44.99          7.78   

   seller_zip_code_prefix seller_city seller_state  
0                    9350        maua           SP  
1                    9350        maua           SP  
2                    9350        maua           SP  
3                    9350        maua           SP  
4                    9350        maua           SP  

[5 rows x 31 columns]

## Insights EDA

In [69]:
# Colunas
for column in df_all.columns:
    print(column)

order_id
customer_id
order_status
order_purchase_timestamp
order_approved_at
order_delivered_carrier_date
order_delivered_customer_date
order_estimated_delivery_date
payment_sequential
payment_type
payment_installments
payment_value
customer_unique_id
customer_zip_code_prefix
customer_city
customer_state
review_id
review_score
review_comment_title
review_comment_message
review_creation_date
review_answer_timestamp
order_item_id
product_id
seller_id
shipping_limit_date
price
freight_value
seller_zip_code_prefix
seller_city
seller_state


In [70]:
# Tirando colunas de id para melhor facilidade
for col in df_all.columns:
    if 'id' in col:
        if col != 'order_id':
            df_all.drop(col, axis=1, inplace=True)
            pass
    else:
        print(col)

order_status
order_purchase_timestamp
order_approved_at
order_delivered_carrier_date
order_delivered_customer_date
order_estimated_delivery_date
payment_sequential
payment_type
payment_installments
payment_value
customer_zip_code_prefix
customer_city
customer_state
review_score
review_comment_title
review_comment_message
review_creation_date
review_answer_timestamp
shipping_limit_date
price
freight_value
seller_zip_code_prefix
seller_city
seller_state


In [72]:
df_all.iloc[:5, :8]

order_id order_status order_purchase_timestamp  \
0  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
1  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
2  e481f51cbdc54678b7cc49136f2d6af7    delivered      2017-10-02 10:56:33   
3  8736140c61ea584cb4250074756d8f3b    delivered      2017-08-10 13:35:55   
4  a0151737f2f0c6c0a5fd69d45f66ceea    delivered      2017-07-12 14:11:58   

     order_approved_at order_delivered_carrier_date  \
0  2017-10-02 11:07:15          2017-10-04 19:55:00   
1  2017-10-02 11:07:15          2017-10-04 19:55:00   
2  2017-10-02 11:07:15          2017-10-04 19:55:00   
3  2017-08-10 13:50:09          2017-08-11 13:52:35   
4  2017-07-12 14:30:18          2017-07-13 12:47:05   

  order_delivered_customer_date order_estimated_delivery_date  \
0           2017-10-10 21:25:13           2017-10-18 00:00:00   
1           2017-10-10 21:25:13           2017-10-18 00:00:00   
2           2017-10-10 21:25:13           2017-10-18 00:00:00   
3           2017-08-16 19:03:36           2017-08-23 00:00:00   
4           2017-07-18 16:24:42           2017-07-25 00:00:00   

   payment_sequential  
0                   1  
1                   3  
2                   2  
3                   1  
4                   1

In [73]:
df_all.iloc[:5, 8:]

payment_type  payment_installments  payment_value  customer_zip_code_prefix  \
0  credit_card                     1          18.12                      3149   
1      voucher                     1           2.00                      3149   
2      voucher                     1          18.59                      3149   
3  credit_card                     1          83.69                      8577   
4  credit_card                     2          52.77                     13820   

     customer_city customer_state  review_score review_comment_title  \
0        sao paulo             SP             4                  NaN   
1        sao paulo             SP             4                  NaN   
2        sao paulo             SP             4                  NaN   
3  itaquaquecetuba             SP             5                  NaN   
4       jaguariuna             SP             4                  NaN   

                              review_comment_message review_creation_date  \
0  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
1  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
2  Não testei o produto ainda, mas ele veio corre...  2017-10-11 00:00:00   
3          custo beneficio, simples de usar e rápido  2017-08-17 00:00:00   
4                                                NaN  2017-07-19 00:00:00   

  review_answer_timestamp  shipping_limit_date  price  freight_value  \
0     2017-10-12 03:43:48  2017-10-06 11:07:15  29.99           8.72   
1     2017-10-12 03:43:48  2017-10-06 11:07:15  29.99           8.72   
2     2017-10-12 03:43:48  2017-10-06 11:07:15  29.99           8.72   
3     2017-08-21 12:43:27  2017-08-16 13:50:09  75.90           7.79   
4     2017-07-24 13:59:59  2017-07-18 14:30:18  44.99           7.78   

   seller_zip_code_prefix seller_city seller_state  
0                    9350        maua           SP  
1                    9350        maua           SP  
2                    9350        maua           SP  
3                    9350        maua           SP  
4                    9350        maua           SP

### Perguntas

    - Contagem de status das ordens;
    - Frequência de compra por mês;
    - Frequência de compra por dia da semana;
    - Tempo médio de aprovação por alguma coisa (approved - purchase time) (hue = a definir);
    - Tempo médio de postagem por alguma coisa (hue = a definir);
    - Tempo médio de entrega ao cliente por alguma coisa (hue = a definir);
    - Diferença entre tempo estimado e tempo de entrega (hue = a definir);
    - Contagem de tipo de pagamento payment_type;
    - Contagem de parcelas payment_installments;
    - Contagem de vendas por estado;
    - Escolher um estado e verificar vendas por cidade;
    - Valor total de vendas por estado (com e sem frete);
    - Escolher um estado e verificar valor total de vendas por cidade;
    - Review score por alguma coisa (hue = a definir), pode ser estado ou tipo do produto;
    - Soma de frete por estado;
    - Contagem de vendas (vendedores/estado)